File: .github/scripts/check-transient-workspace.py
#!/usr/bin/env python3
"""Reject tracked files under the transient repository workspace."""

from __future__ import annotations

import os
from pathlib import Path
import subprocess
import sys


ROOT_ENVIRONMENT_VARIABLE = "TRANSIENT_WORKSPACE_ROOT"


def repository_root() -> Path:
    configured_root = os.environ.get(ROOT_ENVIRONMENT_VARIABLE)
    if configured_root:
        return Path(configured_root)

    result = subprocess.run(
        ["git", "rev-parse", "--show-toplevel"],
        capture_output=True,
        check=False,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(result.stderr.strip() or "unable to resolve the repository root")
    return Path(result.stdout.strip())


def tracked_workspace_files(root: Path) -> list[str]:
    result = subprocess.run(
        ["git", "ls-files", "--", "workspace/"],
        cwd=root,
        capture_output=True,
        check=False,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(result.stderr.strip() or "unable to inspect tracked files")
    return [path for path in result.stdout.splitlines() if path]


def main() -> int:
    try:
        files = tracked_workspace_files(repository_root())
    except (OSError, RuntimeError) as error:
        print(f"Transient workspace check failed: {error}", file=sys.stderr)
        return 1

    if files:
        print("Tracked files under reserved transient workspace/ are not allowed:", file=sys.stderr)
        for path in files:
            print(f"  {path}", file=sys.stderr)
        return 1

    print("No tracked files under reserved transient workspace/.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .github/scripts/check-transient-workspace.mjs
#!/usr/bin/env node
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

export function trackedWorkspaceFiles(root) {
  return execFileSync('git', ['ls-files', '--', 'workspace/'], { cwd: root, encoding: 'utf8' }).split(/\r?\n/).filter(Boolean);
}
function main() {
  let files;
  try {
    const root = path.resolve(process.env.TRANSIENT_WORKSPACE_ROOT ?? execFileSync('git', ['rev-parse', '--show-toplevel'], { encoding: 'utf8' }).trim());
    files = trackedWorkspaceFiles(root);
  } catch (error) { console.error(`Transient workspace check failed: ${error.message}`); return 1; }
  if (files.length) {
    console.error('Tracked files under reserved transient workspace/ are not allowed:');
    for (const file of files) console.error(`  ${file}`);
    return 1;
  }
  console.log('No tracked files under reserved transient workspace/.');
  return 0;
}
if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) process.exitCode = main();
